File: lib_openweather/src/main/java/guepardoapps/library/openweather/controller/NotificationController.java
package guepardoapps.library.openweather.controller;

import android.app.Notification;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.support.annotation.NonNull;

import java.util.Locale;

import guepardoapps.library.openweather.datatransferobjects.NotificationContentDto;
import guepardoapps.library.openweather.utils.BitmapHelper;
import guepardoapps.library.openweather.utils.Logger;

public class NotificationController implements INotificationController {
    private static final String Tag = NotificationController.class.getSimpleName();

    private Context _context;
    private NotificationManager _notificationManager;

    public NotificationController(@NonNull Context context) {
        _context = context;
        _notificationManager = (NotificationManager) _context.getSystemService(Context.NOTIFICATION_SERVICE);
    }

    @Override
    public void CreateNotification(int notificationId, Class<?> receiverActivity, @NonNull NotificationContentDto notificationContent) {
        Logger.getInstance().Debug(Tag, String.format(Locale.getDefault(), "CreateNotification with id %d and notificationContent %s", notificationId, notificationContent));

        Bitmap bitmap = BitmapFactory.decodeResource(_context.getResources(), notificationContent.GetBigIcon());
        bitmap = BitmapHelper.GetCircleBitmap(bitmap, bitmap.getHeight(), bitmap.getWidth());

        Notification.Builder builder = new Notification.Builder(_context);
        builder
                .setSmallIcon(notificationContent.GetIcon())
                .setLargeIcon(bitmap)
                .setContentTitle(notificationContent.GetTitle())
                .setContentText(notificationContent.GetText())
                .setTicker(notificationContent.GetText());

        if (receiverActivity != null) {
            Intent intent = new Intent(_context, receiverActivity);
            PendingIntent pendingIntent = PendingIntent.getActivity(_context, notificationId, intent, 0);
            builder.setContentIntent(pendingIntent);
        }

        Notification notification = builder.build();
        _notificationManager.notify(notificationId, notification);
    }

    @Override
    public void CloseNotification(int notificationId) {
        _notificationManager.cancel(notificationId);
    }
}

File: lib/src/main/java/guepardoapps/lib/openweather/controller/NotificationController.kt
package guepardoapps.lib.openweather.controller

import android.annotation.TargetApi
import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.graphics.BitmapFactory
import android.graphics.Color
import android.os.Build
import android.support.annotation.NonNull
import guepardoapps.lib.openweather.extensions.circleBitmap
import guepardoapps.lib.openweather.models.INotificationContent

class NotificationController(@NonNull private val context: Context) : INotificationController {
    //private val tag: String = NotificationController::class.java.simpleName

    private val channelId: String = "guepardoapps.lib.openweather"
    private val channelName: String = "OpenWeather"
    private val channelDescription: String = "Notifications for open weather library"

    private var notificationManager: NotificationManager? = null

    init {
        notificationManager = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            createNotificationChannel()
        }
    }

    @Suppress("DEPRECATION")
    override fun create(notificationContent: INotificationContent) {
        var bitmap = BitmapFactory.decodeResource(context.resources, notificationContent.largeIcon)
        bitmap = bitmap.circleBitmap(bitmap.height, bitmap.width, Color.BLACK)

        val intent = Intent(context, notificationContent.receiver)
        val pendingIntent = PendingIntent.getActivity(context, 0, intent, PendingIntent.FLAG_UPDATE_CURRENT)

        val notification: Notification
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            notification = Notification.Builder(context, channelId)
                    .setContentTitle(notificationContent.title)
                    .setContentText(notificationContent.text)
                    .setSmallIcon(notificationContent.icon)
                    .setLargeIcon(bitmap)
                    .setChannelId(channelId)
                    .setContentIntent(pendingIntent)
                    .build()
        } else {
            notification = Notification.Builder(context)
                    .setContentTitle(notificationContent.title)
                    .setContentText(notificationContent.text)
                    .setSmallIcon(notificationContent.icon)
                    .setLargeIcon(bitmap)
                    .setContentIntent(pendingIntent)
                    .build()
        }

        notificationManager?.notify(notificationContent.id, notification)
    }

    override fun close() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            notificationManager?.deleteNotificationChannel(channelId)
        } else {

        }
    }

    @TargetApi(Build.VERSION_CODES.O)
    private fun createNotificationChannel() {
        val importance = NotificationManager.IMPORTANCE_LOW
        val channel = NotificationChannel(channelId, channelName, importance)

        channel.description = channelDescription
        channel.enableLights(false)
        //channel.lightColor = Color.GREEN
        channel.enableVibration(false)
        //channel.vibrationPattern = longArrayOf(100, 200, 300, 400, 500, 400, 300, 200, 400)
        notificationManager?.createNotificationChannel(channel)
    }
}